pkg/controller/crdmirroring/crdmirroring_controller.go

```diff
@@ -0,0 +1,213 @@
+		AddFunc:    c.onLegacyCRDAdd,
+		UpdateFunc: c.onLegacyCRDUpdate,
+		DeleteFunc: c.onLegacyCRDDelete,
+	}
+
+	c.informer.Informer().AddEventHandler(handlers)
+	c.listerSycned = c.informer.Informer().HasSynced
+	c.legacyInformer.Informer().AddEventHandler(legacyHandlers)
+	c.legacyListerSynced = c.legacyInformer.Informer().HasSynced
+
+	switch c.CRDName {
+	case NetworkPolicy:
+		c.mirroringHandler = crdhandler.NewNetworkPolicyHandler(c.informer, c.legacyInformer, client, legacyClient, c.CRDName)
+	case ClusterNetworkPolicy:
+		c.mirroringHandler = crdhandler.NewClusterNetworkPolicyHandler(c.informer, c.legacyInformer, client, legacyClient, c.CRDName)
+	case Tier:
+		c.mirroringHandler = crdhandler.NewTierHandler(c.informer, c.legacyInformer, client, legacyClient, c.CRDName)
+	case ClusterGroup:
+		c.mirroringHandler = crdhandler.NewClusterGroupHandler(c.informer, c.legacyInformer, client, legacyClient, c.CRDName)
+	case ExternalEntity:
+		c.mirroringHandler = crdhandler.NewExternalEntityHandler(c.informer, c.legacyInformer, client, legacyClient, c.CRDName)
+	case Traceflow:
+		c.mirroringHandler = crdhandler.NewTraceflowHandler(c.informer, c.legacyInformer, client, legacyClient, c.CRDName)
+	}
+
+	return c
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %vMirroringController", c.CRDName)
+	defer klog.Infof("Shutting down %vMirroringController", c.CRDName)
+
+	if !cache.WaitForNamedCacheSync(fmt.Sprintf("%vMirroringController", c.CRDName), stopCh, c.listerSycned, c.legacyListerSynced) {
+		return
+	}
+
+	klog.Infof("Starting %d worker threads", defaultWorkers)
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, c.workerLoopPeriod, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	cKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(cKey)
+
+	err := c.syncMirroring(cKey.(string))
+	c.handleErr(err, cKey)
+
+	return true
+}
+
+func (c *Controller) syncMirroring(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing for %q legacy CRD. (%v)", key, time.Since(startTime))
+	}()
+
+	klog.V(4).Infof("Sync mirroring CRD (%q)", key)
+	action, from, nn, err := getMirroringInfo(key)
+	if err != nil {
+		c.queue.Forget(key)
+		return nil
+	}
+	namespace, name, err := cache.SplitMetaNamespaceKey(nn)
+	if err != nil {
+		c.queue.Forget(key)
+		return nil
+	}
+	klog.V(4).Infof("Process %s/%s ACTION: %v FROM: %v", namespace, name, action, from)
+
+	switch action {
+	case types.ADD:
+		err = c.mirroringHandler.ADD(from, namespace, name)
+	case types.UPDATE:
+		err = c.mirroringHandler.UPDATE(from, namespace, name)
+	case types.DELETE:
+		err = c.mirroringHandler.DELETE(from, namespace, name)
+	case types.STALE:
+		err = c.mirroringHandler.STALE(namespace, name)
+	}
```

Comment:
There are code that could have been shared are repeated in concrete types. I believe we could reduce code redundancy further by handling more logic here, for example:
```
legacyObj, err := c.mirroringHandler.GetLegacyObject(namespace, name)
if err != nil {
	// Legacy object deleted
	if apierrors.IsNotFound(err) {
		newObj, err := c.mirroringHandler.GetNewObject(namespace, name)
		if err != nil {
			if apierrors.IsNotFound(err) {
				return nil
			}
			return err
		}
		if newObj.GetAnnotations()[types.ManagedBy] == types.ControllerName {
			err := c.mirroringHandler.DeleteNewObject(namespace, name)
			if err != nil {
				return err
			}
		}
		return nil
	}
	return err
}

// Mirror stopped
_, mirrorStopped := legacyObj.GetAnnotations()["crd.antrea.io/stop-mirror"]
if mirrorStopped {
	newObj, err := c.mirroringHandler.GetNewObject(namespace, name)
	if err != nil {
		if apierrors.IsNotFound(err) {
			return nil
		}
		return err
	}
	if newObj.GetAnnotations()[types.ManagedBy] != types.ControllerName {
		return nil
	}
	toUpdate := newObj.DeepCopy()
	delete(toUpdate.GetAnnotations(), types.ManagedBy)
	_, err := c.mirroringHandler.UpdateNewObject(namespace, name)
	if err != nil {
		return err
	}
	return nil
}

// Mirror
newObj, err := c.mirroringHandler.GetNewObject(namespace, name)
if err != nil {
	if apierrors.IsNotFound(err) {
		_, err := c.mirroringHandler.CreateNewObject(legacyObj)
		if err != nil {
			return err
		}
		return nil
	}
	return err
}
err = c.mirroringHandler.SyncObjects(legacyObj, newObj)
if err != nil {
	return err
}
return nil
```

Then we only need the concrete handler to implement their specific CRUD methods: GetLegacyObject, GetNewObject, DeleteNewObject, UpdateNewObject, CreateNewObject, SyncObjects. 

The methods should use metav1.Object as parameter and return types to be generic.

Reply:
Thanks, this helps a lot.